### PR TITLE
Remove need for running in privileged mode

### LINKED ISSUE #2: Asymmetric routing / read only filesyste errors; leads to container stopping
**OS & Kernel:**
`Linux hostname 5.3.0-51-generic #44-Ubuntu SMP Wed Apr 22 21:09:44 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Behavior:** 
Newly built docker image yields the following error _unless_ it is run in privileged mode:
`sysctl: setting key "net.ipv4.conf.all.src_valid_mark": Read-only file system`

This behavior persists even after changing the relevant values in /etc/sysctl.conf as described in the README and rebooting. This behavior only occurs for _newly_ generated images; running the jordanpotter/wireguard docker image obtained from the docker registry works fine with no errors under the same system.

**Command used to build image:**
`docker build -t jordanpotter/wireguard .`

Run in the docker-wireguard directory.

**Relevant wireguard conf:**
I tested this using a simple wireguard config from mullvad:

```

[Interface]
PrivateKey = xxxxxxxx
Address = xxxxxx
DNS = xxxxxx

[Peer]
PublicKey = xxxxx
AllowedIPs = xxxxxx
Endpoint = xxxxxx

```

### PR DESCRIPTION
**Pull request by bobbypage**
To remove the need for running privileged mode, set the necessary
`net.ipv4.conf.all.src_valid_mark=1` sysctl via docker and modify the
`/usr/bin/wg-quick` script from setting the sysctl

* Set `net.ipv4.conf.all.src_valid_mark=1` sysctl via docker `--sysctl`

* Set CAPs of NET_ADMIN and SYS_MODULE instead of using`--privileged`.

* Check that `net.ipv4.conf.all.src_valid_mark=1` is set in entrypoint
  script

Fixes #2

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -11,15 +11,25 @@ fi
 config=`echo $configs | head -n 1`
 interface="${config%.*}"
 
+if [[ "$(cat /proc/sys/net/ipv4/conf/all/src_valid_mark)" != "1" ]]; then
+    echo "sysctl net.ipv4.conf.all.src_valid_mark=1 is not set" >&2
+    exit 1
+fi
+
+sed -i "s:sysctl -q net.ipv4.conf.all.src_valid_mark=1:echo skipping setting net.ipv4.conf.all.src_valid_mark:" /usr/bin/wg-quick
```
Comment (jordanpotter) [inline entrypoint.sh:19]: Just to double-check, would you mind responding to this comment with the before and after of `wg-quick`?

File: entrypoint.sh [hunk 2]
```diff
@@ -11,15 +11,25 @@ fi
...
 
 docker6_network="$(ip -o addr show dev eth0 | awk '$3 == "inet6" {print $4}')"
-docker6_network_rule=$([ ! -z "$docker6_network" ] && echo "! -d $docker6_network" || echo "")
-ip6tables -I OUTPUT ! -o $interface -m mark ! --mark $(wg show $interface fwmark) -m addrtype ! --dst-type LOCAL $docker6_network_rule -j REJECT
+if [[ -z "${docker6_network}" ]]; then
+    echo "skipping ipv6 killswitch setup since ipv6 interface was not found..." >&2
+else
+    docker6_network_rule=$([ ! -z "$docker6_network" ] && echo "! -d $docker6_network" || echo "")
+    ip6tables -I OUTPUT ! -o $interface -m mark ! --mark $(wg show $interface fwmark) -m addrtype ! --dst-type LOCAL $docker6_network_rule -j REJECT
+fi
```
Comment (jordanpotter) [inline entrypoint.sh:32]: Could you elaborate on why this change was required? It's a solid change, but I don't quite follow how it's related to removing `--privileged`

File: entrypoint.sh [hunk 3]
```diff
@@ -11,15 +11,25 @@ fi
...
 
 docker_network="$(ip -o addr show dev eth0 | awk '$3 == "inet" {print $4}')"
 docker_network_rule=$([ ! -z "$docker_network" ] && echo "! -d $docker_network" || echo "")
 iptables -I OUTPUT ! -o $interface -m mark ! --mark $(wg show $interface fwmark) -m addrtype ! --dst-type LOCAL $docker_network_rule -j REJECT
 
 docker6_network="$(ip -o addr show dev eth0 | awk '$3 == "inet6" {print $4}')"
-docker6_network_rule=$([ ! -z "$docker6_network" ] && echo "! -d $docker6_network" || echo "")
-ip6tables -I OUTPUT ! -o $interface -m mark ! --mark $(wg show $interface fwmark) -m addrtype ! --dst-type LOCAL $docker6_network_rule -j REJECT
+if [[ -z "${docker6_network}" ]]; then
+    echo "skipping ipv6 killswitch setup since ipv6 interface was not found..." >&2
```
Comment (jordanpotter) [inline entrypoint.sh:28]: Mind starting the output with a capital letter to be consistent with the style on line 7?

```suggestion
    echo "Skipping ipv6 killswitch setup since ipv6 interface was not found" >&2
```

Or optionally change line 7 to be consistent with this one. Either way 😄

**Review by jordanpotter — COMMENTED** [review]
@bobbypage this is a brilliant PR! Thank you for spending the time to make this better!

**Comment by jordanpotter** [comment]
I just tested this on a machine running Debian buster with kernel `4.19.0-9-amd64`.

Saw the following error:
```
$ docker run --rm --cap-add NET_ADMIN --cap-add SYS_MODULE --sysctl net.ipv4.conf.all.src_valid_mark=1 -v /etc/wireguard/mullvad.conf:/etc/wireguard/mullvad.conf docker-wireguard
[#] ip link add mullvad type wireguard
[#] wg setconf mullvad /dev/fd/63
[#] ip -4 address add <redacted>/32 dev mullvad
[#] ip link set mtu 1420 up dev mullvad
[#] resolvconf -a mullvad -m 0 -x
[#] wg set mullvad fwmark 51820
[#] ip -4 route add 0.0.0.0/0 dev mullvad table 51820
[#] ip -4 rule add not fwmark 51820 table 51820
[#] ip -4 rule add table main suppress_prefixlength 0
[#] sysctl -q net.ipv4.conf.all.src_valid_mark=1
sysctl: setting key "net.ipv4.conf.all.src_valid_mark": Read-only file system
[#] resolvconf -d mullvad -f
[#] ip -4 rule delete table 51820
[#] ip -4 rule delete table main suppress_prefixlength 0
[#] ip link delete dev mullvad
```

File: entrypoint.sh (hunk 1 above)
Comment (bobbypage) [reply entrypoint.sh:19]: Before: https://termbin.com/926v
After: https://termbin.com/rcod

Diff:
```
@@ -237,7 +237,7 @@
        printf -v restore '%sCOMMIT\n*mangle\n-I POSTROUTING -m mark --mark %d -p udp -j CONNMARK --save-mark %s\n-I PREROUTING -p udp -j CONNMARK --restore-mark %s\nCOMMIT\n' "$restore" $table "$marker" "$marker"
        printf -v nftcmd '%sadd rule %s %s postmangle meta l4proto udp mark %d ct mark set mark \n' "$nftcmd" "$pf" "$nftable" $table
        printf -v nftcmd '%sadd rule %s %s premangle meta l4proto udp meta mark set ct mark \n' "$nftcmd" "$pf" "$nftable"
-       [[ $proto == -4 ]] && cmd sysctl -q net.ipv4.conf.all.src_valid_mark=1
+       [[ $proto == -4 ]] && cmd echo skipping setting net.ipv4.conf.all.src_valid_mark
        if type -p nft >/dev/null; then
                cmd nft -f <(echo -n "$nftcmd")
        else
```

File: entrypoint.sh (hunk 2 above)
Comment (bobbypage) [reply entrypoint.sh:32]: Sorry for the confusion, this is a separate change from the privileged. I was having issues starting the container due to not having a docker ipv6 network setup, so I added this to prevent setting up ipv6 rules if there isn't an ipv6 interface.

File: entrypoint.sh (hunk 3 above)
Comment (bobbypage) [reply entrypoint.sh]: Done.

File: entrypoint.sh (hunk 2 above)
Comment (jordanpotter) [reply entrypoint.sh]: Ah, fair enough! In that case, would you mind making this a separate PR?

**Comment by jordanpotter** [comment]
I just tested this on a machine running Debian buster with kernel `4.19.0-9-amd64`.

```
$ docker run --rm --name docker-wireguard --cap-add NET_ADMIN --cap-add SYS_MODULE --sysctl net.ipv4.conf.all.src_valid_mark=1 -v /etc/wireguard/mullvad.conf:/etc/wireguard/mullvad.conf docker-wireguard
[#] ip link add mullvad type wireguard
[#] wg setconf mullvad /dev/fd/63
[#] ip -4 address add <redacted>/32 dev mullvad
[#] ip link set mtu 1420 up dev mullvad
[#] resolvconf -a mullvad -m 0 -x
[#] wg set mullvad fwmark 51820
[#] ip -4 route add 0.0.0.0/0 dev mullvad table 51820
[#] ip -4 rule add not fwmark 51820 table 51820
[#] ip -4 rule add table main suppress_prefixlength 0
skipping setting net.ipv4.conf.all.src_valid_mark
[#] echo skipping setting net.ipv4.conf.all.src_valid_mark
[#] iptables-restore -n
skipping ipv6 killswitch setup since ipv6 interface was not found...
```

Then:
```
$ docker run -it --rm --net=container:docker-wireguard-pr appropriate/curl http://httpbin.org/ip
curl: (6) Could not resolve host: httpbin.org
```

Do you get a different result on your machine?

File: entrypoint.sh (hunk 1 above)
Comment (jordanpotter) [reply entrypoint.sh:19]: Perfect! Thanks!

**Comment by bobbypage** [comment]
Huh, that's strange it works on my side. I'm testing this on Ubuntu 20.04.1 LTS, kernel `5.4.0-42-generic`

```
$ sudo docker build -t bobbypage/docker-wireguard:latest .
Sending build context to Docker daemon    171kB
Step 1/4 : FROM ubuntu:20.04
 ---> 4e2eef94cd6b
Step 2/4 : RUN apt-get update && apt-get install -y openresolv iptables iproute2 wireguard
 ---> Using cache
 ---> f395a406a71c
Step 3/4 : COPY entrypoint.sh /entrypoint.sh
 ---> b0e63f5f4527
Step 4/4 : ENTRYPOINT ["/entrypoint.sh"]
 ---> Running in 94dac4dd08be
Removing intermediate container 94dac4dd08be
 ---> 8362dcf93eaf
Successfully built 8362dcf93eaf
Successfully tagged bobbypage/docker-wireguard:latest
$ sudo docker run --name wireguard --cap-add NET_ADMIN --cap-add SYS_MODULE --sysctl net.ipv4.conf.all.src_valid_mark=1 -v $(pwd)/config/mullvad/mullvad-ca10-ipv4.conf:/etc/wireguard/mullvad.conf bobbypage/docker-wireguard:latest
```

```
$ sudo docker run --name wireguard --cap-add NET_ADMIN --cap-add SYS_MODULE --sysctl net.ipv4.conf.all.src_valid_mark=1 -v $(pwd)/config/mullvad/mullvad-ca10-ipv4.conf:/etc/wireguard/mullvad.conf bobbypage/docker-wireguard:latest
[#] ip link add mullvad type wireguard
[#] wg setconf mullvad /dev/fd/63
[#] ip -4 address add <redacted> dev mullvad
[#] ip link set mtu 1420 up dev mullvad
[#] resolvconf -a mullvad -m 0 -x
[#] wg set mullvad fwmark 51820
[#] ip -4 route add 0.0.0.0/0 dev mullvad table 51820
[#] ip -4 rule add not fwmark 51820 table 51820
[#] ip -4 rule add table main suppress_prefixlength 0
[#] echo skipping setting net.ipv4.conf.all.src_valid_mark
skipping setting net.ipv4.conf.all.src_valid_mark
[#] iptables-restore -n
```

```
$ sudo docker run -it --rm --net=container:wireguard appropriate/curl http://httpbin.org/ip
{
  "origin": "<SERVER_IP>"
}
```

**Comment by jordanpotter** [comment]
Hrmm, give me a few minutes to check my wireguard config... 🤔

File: entrypoint.sh (hunk 2 above)
Comment (bobbypage) [reply entrypoint.sh]: Sure, removed it from this PR and opened a new PR: https://github.com/jordanpotter/docker-wireguard/pull/4

Comment (jordanpotter) [reply entrypoint.sh]: Thank you so much!

**Comment by jordanpotter** [comment]
I was mistaken! The wireguard configuration I was using for testing was being used by another wireguard instance, which caused the conflict. We're good now 👍

**Review by jordanpotter — APPROVED** [review]
Thank you again @bobbypage for the wonderful PR!
